### ISSUE TITLE
Swiss table containers fail tests with unsigned characters on GCC

### ISSUE THREAD

**Issue by bocon13** [abseil/abseil-cpp#209]
When Abseil is built with `-funsigned-char` on GCC, containers like `flat_hash_map` return non-deterministic results when attempting to find or iterate over all elements that have been emplaced. Sometimes elements that have just been added cannot be found; frequently the iterator skips all or most elements.

Environment:
Ubuntu 16.04.5 (xenial) 
Kernel: 4.4.0-137-generic
Compiler: gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.10)
Bazel: 0.18.0
Abseil: commit a4c3ffff11eec0ee45742f915c255e9f870b7e0f

To reproduce the issue:
```
bazel test --cxxopt='-funsigned-char' //...
<snip>
//absl/container:flat_hash_map_test                                      FAILED in 0.3s
//absl/container:flat_hash_set_test                                      FAILED in 0.2s
//absl/container:node_hash_map_test                                      FAILED in 0.3s
//absl/container:node_hash_set_test                                      FAILED in 0.3s
//absl/container:raw_hash_set_allocator_test                             FAILED in 0.1s
//absl/container:raw_hash_set_test                                       FAILED in 0.9s

Executed 6 out of 121 tests: 115 tests pass and 6 fail locally.
```

Note: This does not appear to be an issue when building with Clang on Mac OS X.

**Comment by fowles** [comment]
in debug builds swisstable has extra asserts that will help track this down

**Comment by derekmauro** [comment]
The tests pass with `bazel test --cxxopt='-funsigned-char' --copt=-DSWISSTABLE_HAVE_SSE2=0 --copt=-DSWISSTABLE_HAVE_SSSE3=0 absl/container/...` so I suspect the issue is in the use of SSE instructions.

**Comment by derekmauro** [comment]
I *think* this is a GCC bug.

The implementation using SSE2 relies on `_mm_cmpgt_epi8` in some places, for example:
https://github.com/abseil/abseil-cpp/blob/cc8dcd307b76a575d2e3e0958a4fe4c7193c2f68/absl/container/internal/raw_hash_set.h#L375

GCC's implementation of `_mm_cmpgt_epi8`:
https://github.com/gcc-mirror/gcc/blob/913b81c4d40ca6e1b157ea5785376d8de58b42fc/gcc/config/i386/emmintrin.h#L1331-L1335

and `__v16qi`:
https://github.com/gcc-mirror/gcc/blob/913b81c4d40ca6e1b157ea5785376d8de58b42fc/gcc/config/i386/emmintrin.h#L47
I think a type with `signed char` instead of `char` is needed to implement `_mm_cmpgt_epi8` correctly since `_mm_cmpgt_epi8` is supposed to compare signed values.

Clang seems to get this right:
https://github.com/llvm-mirror/clang/blob/5fd1ab66bf97297c2b71ce3ec2bb33d579bca06e/lib/Headers/emmintrin.h#L3247-L3253

**Comment by derekmauro** [comment]
I've reported this to GCC: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87853

**Comment by derekmauro** [comment]
f95179062eb65ce40895cc76f1398cce25394369 contains a workaround for this. It should be working now, but you will be sacrificing some performance if you use GCC with -funsigned-char. I'm not sure if there is anything else we can do besides maybe working around the GCC bug by writing some assembly.

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87853 already has a proposed patch for the GCC bug, but you'll obviously have to update your compiler and will we have to change our code to only use the workaround on known buggy GCC versions.

**Comment by sneves** [comment]
Instead of reverting to the generic implementation, you may also consider using an alternative signed comparison implementation that avoids the offending instruction:

```
static inline __m128i _mm_cmpgt_epi8_alternative(__m128i a, __m128i b) {
	const __m128i mask = _mm_set1_epi8(0x80);
	const __m128i diff = _mm_subs_epi8(b, a);
	return _mm_cmpeq_epi8(_mm_and_si128(diff, mask), mask);
}
```

**Comment by derekmauro** [comment]
@sneves Thank you for the suggestion. This is done in 7990fd459e9339467814ddb95000c87cb1e4d945.

**Comment by bocon13** [comment]
Thanks, @derekmauro, for the fixes and upstream report, and @sneves, for the suggestions on maintaining performance on GCC by continuing to use the SSE instruction albeit correctly! :)

I've confirmed that this issue is resolved in my environment on 7990fd459e9339467814ddb95000c87cb1e4d945, so I'm closing this issue.
